{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Extracting per-read per-position dwell-times from fast5 files (notebook 2)\n",
    "\n",
    "The last notebook was getting too long. We located the data, let's try to extract it in this notebook."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import numpy.lib.recfunctions as rfn # tools for working with structured arrays\n",
    "import h5py # standard Python library for working with HDF5 files - http://docs.h5py.org/en/stable/quick.html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "fast5_path = 'data/01936724-8ae4-4a0f-99ce-c67f3a334c1f.fast5'\n",
    "reference_fasta_path = 'data/RNA_section__454_9627.fa'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "group: <HDF5 file \"01936724-8ae4-4a0f-99ce-c67f3a334c1f.fast5\" (mode r)>\n",
      "attribute values:\n",
      "attribute: file_version\n",
      "value: 2.0\n",
      "datasets: []\n",
      "subgroups: [u'Analyses', u'Raw', u'UniqueGlobalKey']\n",
      "\n",
      "\tgroup: <HDF5 group \"/Analyses\" (3 members)>\n",
      "\tattribute values:\n",
      "\tdatasets: []\n",
      "\tsubgroups: [u'Segmentation_000', u'Basecall_1D_000', u'RawGenomeCorrected_000']\n",
      "\n",
      "\t\tgroup: <HDF5 group \"/Analyses/Segmentation_000\" (1 members)>\n",
      "\t\tattribute values:\n",
      "\t\tattribute: name\n",
      "\t\tvalue: MinKNOW-Live-Basecalling\n",
      "\t\tattribute: version\n",
      "\t\tvalue: 3.6.0\n",
      "\t\tattribute: time_stamp\n",
      "\t\tvalue: 2020-01-08T12:27:08Z\n",
      "\t\tdatasets: []\n",
      "\t\tsubgroups: [u'Summary']\n",
      "\n",
      "\t\t\tgroup: <HDF5 group \"/Analyses/Segmentation_000/Summary\" (1 members)>\n",
      "\t\t\tattribute values:\n",
      "\t\t\tattribute: return_status\n",
      "\t\t\tvalue: Workflow successful\n",
      "\t\t\tdatasets: []\n",
      "\t\t\tsubgroups: [u'segmentation']\n",
      "\n",
      "\t\t\t\tgroup: <HDF5 group \"/Analyses/Segmentation_000/Summary/segmentation\" (0 members)>\n",
      "\t\t\t\tattribute values:\n",
      "\t\t\t\tattribute: duration_template\n",
      "\t\t\t\tvalue: 409005\n",
      "\t\t\t\tattribute: first_sample_template\n",
      "\t\t\t\tvalue: 5588\n",
      "\t\t\t\tattribute: has_template\n",
      "\t\t\t\tvalue: 1\n",
      "\t\t\t\tdatasets: []\n",
      "\t\t\t\tsubgroups: []\n",
      "\n",
      "\t\tgroup: <HDF5 group \"/Analyses/Basecall_1D_000\" (1 members)>\n",
      "\t\tattribute values:\n",
      "\t\tdatasets: []\n",
      "\t\tsubgroups: [u'BaseCalled_template']\n",
      "\n",
      "\t\t\tgroup: <HDF5 group \"/Analyses/Basecall_1D_000/BaseCalled_template\" (1 members)>\n",
      "\t\t\tattribute values:\n",
      "\t\t\tdatasets: [u'Fastq']\n",
      "\t\t\tsubgroups: []\n",
      "\n",
      "\t\tgroup: <HDF5 group \"/Analyses/RawGenomeCorrected_000\" (1 members)>\n",
      "\t\tattribute values:\n",
      "\t\tattribute: tombo_version\n",
      "\t\tvalue: 1.5.1\n",
      "\t\tattribute: basecall_group\n",
      "\t\tvalue: Basecall_1D_000\n",
      "\t\tdatasets: []\n",
      "\t\tsubgroups: [u'BaseCalled_template']\n",
      "\n",
      "\t\t\tgroup: <HDF5 group \"/Analyses/RawGenomeCorrected_000/BaseCalled_template\" (2 members)>\n",
      "\t\t\tattribute values:\n",
      "\t\t\tattribute: status\n",
      "\t\t\tvalue: success\n",
      "\t\t\tattribute: rna\n",
      "\t\t\tvalue: True\n",
      "\t\t\tattribute: signal_match_score\n",
      "\t\t\tvalue: 1.28494354458\n",
      "\t\t\tattribute: shift\n",
      "\t\t\tvalue: 570.098736793\n",
      "\t\t\tattribute: scale\n",
      "\t\t\tvalue: 73.7784909873\n",
      "\t\t\tattribute: norm_type\n",
      "\t\t\tvalue: median\n",
      "\t\t\tattribute: lower_lim\n",
      "\t\t\tvalue: -5.0\n",
      "\t\t\tattribute: upper_lim\n",
      "\t\t\tvalue: 5.0\n",
      "\t\t\tattribute: outlier_threshold\n",
      "\t\t\tvalue: 5.0\n",
      "\t\t\tdatasets: [u'Events']\n",
      "\t\t\tsubgroups: [u'Alignment']\n",
      "\n",
      "\t\t\t\tgroup: <HDF5 group \"/Analyses/RawGenomeCorrected_000/BaseCalled_template/Alignment\" (0 members)>\n",
      "\t\t\t\tattribute values:\n",
      "\t\t\t\tattribute: mapped_start\n",
      "\t\t\t\tvalue: 33\n",
      "\t\t\t\tattribute: mapped_end\n",
      "\t\t\t\tvalue: 9166\n",
      "\t\t\t\tattribute: mapped_strand\n",
      "\t\t\t\tvalue: +\n",
      "\t\t\t\tattribute: mapped_chrom\n",
      "\t\t\t\tvalue: truncated_hiv_rna_genome\n",
      "\t\t\t\tattribute: clipped_bases_start\n",
      "\t\t\t\tvalue: 18\n",
      "\t\t\t\tattribute: clipped_bases_end\n",
      "\t\t\t\tvalue: 13\n",
      "\t\t\t\tattribute: num_insertions\n",
      "\t\t\t\tvalue: 380\n",
      "\t\t\t\tattribute: num_deletions\n",
      "\t\t\t\tvalue: 459\n",
      "\t\t\t\tattribute: num_matches\n",
      "\t\t\t\tvalue: 8302\n",
      "\t\t\t\tattribute: num_mismatches\n",
      "\t\t\t\tvalue: 372\n",
      "\t\t\t\tdatasets: []\n",
      "\t\t\t\tsubgroups: []\n",
      "\n",
      "\tgroup: <HDF5 group \"/Raw\" (1 members)>\n",
      "\tattribute values:\n",
      "\tdatasets: []\n",
      "\tsubgroups: [u'Reads']\n",
      "\n",
      "\t\tgroup: <HDF5 group \"/Raw/Reads\" (1 members)>\n",
      "\t\tattribute values:\n",
      "\t\tdatasets: []\n",
      "\t\tsubgroups: [u'Read_28507']\n",
      "\n",
      "\t\t\tgroup: <HDF5 group \"/Raw/Reads/Read_28507\" (1 members)>\n",
      "\t\t\tattribute values:\n",
      "\t\t\tattribute: start_time\n",
      "\t\t\tvalue: 152704672\n",
      "\t\t\tattribute: duration\n",
      "\t\t\tvalue: 414593\n",
      "\t\t\tattribute: read_number\n",
      "\t\t\tvalue: 28507\n",
      "\t\t\tattribute: start_mux\n",
      "\t\t\tvalue: 4\n",
      "\t\t\tattribute: read_id\n",
      "\t\t\tvalue: 01936724-8ae4-4a0f-99ce-c67f3a334c1f\n",
      "\t\t\tattribute: median_before\n",
      "\t\t\tvalue: 213.109191895\n",
      "\t\t\tattribute: end_reason\n",
      "\t\t\tvalue: 4\n",
      "\t\t\tdatasets: [u'Signal']\n",
      "\t\t\tsubgroups: []\n",
      "\n",
      "\tgroup: <HDF5 group \"/UniqueGlobalKey\" (3 members)>\n",
      "\tattribute values:\n",
      "\tdatasets: []\n",
      "\tsubgroups: [u'channel_id', u'context_tags', u'tracking_id']\n",
      "\n",
      "\t\tgroup: <HDF5 group \"/UniqueGlobalKey/channel_id\" (0 members)>\n",
      "\t\tattribute values:\n",
      "\t\tattribute: digitisation\n",
      "\t\tvalue: 8192.0\n",
      "\t\tattribute: offset\n",
      "\t\tvalue: 1.0\n",
      "\t\tattribute: range\n",
      "\t\tvalue: 1179.47839355\n",
      "\t\tattribute: sampling_rate\n",
      "\t\tvalue: 3012.0\n",
      "\t\tattribute: channel_number\n",
      "\t\tvalue: 475\n",
      "\t\tdatasets: []\n",
      "\t\tsubgroups: []\n",
      "\n",
      "\t\tgroup: <HDF5 group \"/UniqueGlobalKey/context_tags\" (0 members)>\n",
      "\t\tattribute values:\n",
      "\t\tattribute: barcoding_enabled\n",
      "\t\tvalue: 0\n",
      "\t\tattribute: basecall_config_filename\n",
      "\t\tvalue: rna_r9.4.1_70bps_fast.cfg\n",
      "\t\tattribute: experiment_duration_set\n",
      "\t\tvalue: 4320\n",
      "\t\tattribute: experiment_type\n",
      "\t\tvalue: rna\n",
      "\t\tattribute: local_basecalling\n",
      "\t\tvalue: 1\n",
      "\t\tattribute: package\n",
      "\t\tvalue: bream4\n",
      "\t\tattribute: package_version\n",
      "\t\tvalue: 4.3.12\n",
      "\t\tattribute: sample_frequency\n",
      "\t\tvalue: 3012\n",
      "\t\tattribute: sequencing_kit\n",
      "\t\tvalue: sqk-rna002\n",
      "\t\tdatasets: []\n",
      "\t\tsubgroups: []\n",
      "\n",
      "\t\tgroup: <HDF5 group \"/UniqueGlobalKey/tracking_id\" (0 members)>\n",
      "\t\tattribute values:\n",
      "\t\tattribute: asic_id\n",
      "\t\tvalue: 482514390\n",
      "\t\tattribute: asic_id_eeprom\n",
      "\t\tvalue: 5284531\n",
      "\t\tattribute: asic_temp\n",
      "\t\tvalue: 22.303856\n",
      "\t\tattribute: asic_version\n",
      "\t\tvalue: IA02D\n",
      "\t\tattribute: auto_update\n",
      "\t\tvalue: 1\n",
      "\t\tattribute: auto_update_source\n",
      "\t\tvalue: https://mirror.oxfordnanoportal.com/software/MinKNOW/\n",
      "\t\tattribute: bream_is_standard\n",
      "\t\tvalue: 0\n",
      "\t\tattribute: configuration_version\n",
      "\t\tvalue: 1.0.1\n",
      "\t\tattribute: device_id\n",
      "\t\tvalue: MN29998\n",
      "\t\tattribute: device_type\n",
      "\t\tvalue: minion\n",
      "\t\tattribute: distribution_status\n",
      "\t\tvalue: stable\n",
      "\t\tattribute: distribution_version\n",
      "\t\tvalue: 19.12.2\n",
      "\t\tattribute: exp_script_name\n",
      "\t\tvalue: sequencing\\sequencing_MIN106_MIN107_RNA:FLO-MIN106:SQK-RNA002:True\n",
      "\t\tattribute: exp_script_purpose\n",
      "\t\tvalue: sequencing_run\n",
      "\t\tattribute: exp_start_time\n",
      "\t\tvalue: 2020-01-07T22:19:44Z\n",
      "\t\tattribute: flow_cell_id\n",
      "\t\tvalue: FAN05606\n",
      "\t\tattribute: flow_cell_product_code\n",
      "\t\tvalue: FLO-MIN106\n",
      "\t\tattribute: guppy_version\n",
      "\t\tvalue: 3.2.8+bd67289c\n",
      "\t\tattribute: heatsink_temp\n",
      "\t\tvalue: 32.472656\n",
      "\t\tattribute: hostname\n",
      "\t\tvalue: GLEE-2DS9H03\n",
      "\t\tattribute: installation_type\n",
      "\t\tvalue: nc\n",
      "\t\tattribute: local_firmware_file\n",
      "\t\tvalue: 1\n",
      "\t\tattribute: operating_system\n",
      "\t\tvalue: Windows 10.0\n",
      "\t\tattribute: protocol_group_id\n",
      "\t\tvalue: 010720_HIVRNA_SSIV_multioligos\n",
      "\t\tattribute: protocol_run_id\n",
      "\t\tvalue: 85e3499c-195d-4d30-9f38-ec67c245367d\n",
      "\t\tattribute: protocols_version\n",
      "\t\tvalue: 4.3.12\n",
      "\t\tattribute: run_id\n",
      "\t\tvalue: 0bc4549742d8f2e47ef1c96ae611fbfe6c1c6847\n",
      "\t\tattribute: sample_id\n",
      "\t\tvalue: 010720_HIVRNA_SSIV_multioligos\n",
      "\t\tattribute: usb_config\n",
      "\t\tvalue: MinION_fx3_1.1.1_ONT#MinION_fpga_1.1.0#bulk#Auto\n",
      "\t\tattribute: version\n",
      "\t\tvalue: 3.6.0\n",
      "\t\tdatasets: []\n",
      "\t\tsubgroups: []\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Chris Viola wrote this nice code to recursively describe the whole HDF5 file structure\n",
    "\n",
    "def explore_hdf5_file(hdf5_group,tab_str):\n",
    "    '''\n",
    "    Given an HDF5 root key, hdf5_group prints info on its name, attributes, and subgroups. Does so \n",
    "    for the entire tree recursively. \n",
    "    '''\n",
    "    assert isinstance(hdf5_group, h5py._hl.group.Group), 'Error: This is not an hdf5 group. Its type is {}'.format(type(hdf5_group))\n",
    "    \n",
    "    print(tab_str + 'group: {}'.format(hdf5_group))\n",
    "    \n",
    "    print(tab_str + 'attribute values:')\n",
    "    for key in hdf5_group.attrs.keys():\n",
    "        print(tab_str + 'attribute: {}'.format(key)) #attr name \n",
    "        print(tab_str + 'value: {}'.format(hdf5_group.attrs[key])) #value of attr\n",
    "    \n",
    "    datasets_and_subgroups = hdf5_group.keys()\n",
    "    datasets = [x for x in datasets_and_subgroups if isinstance(hdf5_group[x], h5py._hl.dataset.Dataset)]\n",
    "    subgroups = [x for x in datasets_and_subgroups if isinstance(hdf5_group[x], h5py._hl.group.Group)]\n",
    "    print(tab_str + 'datasets: {}'.format(datasets))\n",
    "    print(tab_str + 'subgroups: {}'.format(subgroups))\n",
    "    print('')\n",
    "    tab_str+='\\t'\n",
    "    if len(subgroups) != 0:\n",
    "        for subgroup in subgroups:\n",
    "            explore_hdf5_file(hdf5_group[subgroup],tab_str)\n",
    "            \n",
    "with h5py.File(fast5_path, 'r') as file:\n",
    "    explore_hdf5_file(file,'')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "One insight from looking at all the attributes in the file:\n",
    "it appears attribute `sample_frequency` of `/UniqueGlobalKey/context_tags` tells us the sampling frequency in Hz.\n",
    "(I have not confirmed this!!)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# store sample frequency for later use\n",
    "with h5py.File(fast5_path, 'r') as file:\n",
    "    sample_frequency = file['/UniqueGlobalKey/context_tags'].attrs['sample_frequency']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'h5py._hl.dataset.Dataset'>\n",
      "<HDF5 dataset \"Events\": shape (9133,), type \"|V25\">\n",
      "[('norm_mean', '<f8'), ('norm_stdev', '<f8'), ('start', '<u4'), ('length', '<u4'), ('base', 'S1')]\n",
      "[[(2.1117437 , nan,      0, 20, 'G')]\n",
      " [(2.31800499, nan,     20, 17, 'A')]\n",
      " [(1.93237796, nan,     37,  6, 'G')]\n",
      " ...\n",
      " [(1.66882629, nan, 405536, 27, 'G')]\n",
      " [(2.28478419, nan, 405563, 12, 'A')]\n",
      " [(0.71333169, nan, 405575, 11, 'G')]]\n"
     ]
    }
   ],
   "source": [
    "# look at Events dataset\n",
    "with h5py.File(fast5_path, 'r') as file:\n",
    "    events_dset = file['/Analyses/RawGenomeCorrected_000/BaseCalled_template/Events']\n",
    "    print(type(events_dset))\n",
    "    print(events_dset)\n",
    "    print(events_dset.dtype)\n",
    "    print(events_dset[()].reshape(-1,1)) # reshape function is just to make rows line up in a column"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[('norm_mean', '<f8'), ('norm_stdev', '<f8'), ('start', '<u4'), ('length', '<u4'), ('base', 'S1')]\n",
      "[[(2.1117437 , nan,      0, 20, 'G')]\n",
      " [(2.31800499, nan,     20, 17, 'A')]\n",
      " [(1.93237796, nan,     37,  6, 'G')]\n",
      " ...\n",
      " [(1.66882629, nan, 405536, 27, 'G')]\n",
      " [(2.28478419, nan, 405563, 12, 'A')]\n",
      " [(0.71333169, nan, 405575, 11, 'G')]]\n"
     ]
    }
   ],
   "source": [
    "# convert Events dataset to numpy array\n",
    "with h5py.File(fast5_path, 'r') as file:\n",
    "    events_dset = file['/Analyses/RawGenomeCorrected_000/BaseCalled_template/Events']\n",
    "    events_array = np.asarray(events_dset)\n",
    "    \n",
    "print(events_array.dtype)\n",
    "print(events_array.reshape(-1,1)) # reshape function is just to make rows line up in a column"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The sequence from `events_array` is indeed a subsequence of the reference genome\n"
     ]
    }
   ],
   "source": [
    "# double-check that the genetic sequence in events_array is a subsequence of the reference genome\n",
    "\n",
    "# Here's the sequence from the events table\n",
    "events_sequence = ''.join(events_array['base']) # concatenate letters from \"length\" column of table\n",
    "\n",
    "# Here's the reference genome\n",
    "with open(reference_fasta_path, 'rt') as file:\n",
    "    # For some reason, file.read() outputs a string including \"\\r\\n\". That's supposed to be a newline separator\n",
    "    reference_genome = file.read().split('\\r\\n')[1] # split by the weird newline separator, then access the second line with \"[1]\"\n",
    "\n",
    "# This method would need modification to work on more sophisticated reference genomes with\n",
    "# multiple isoforms, multiple chromosomes, multiple strands, etc.\n",
    "assert events_sequence.upper() in reference_genome.upper() # one of these strings is lowercase, the other upper, so we must convert\n",
    "print('The sequence from `events_array` is indeed a subsequence of the reference genome')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's just create an array called `pos_array` (of the same shape as `events_array`) that describes genomic position.\n",
    "For each entry in `events_array`, the corresponding entry in `pos_array` will be\n",
    "the position of the corresponding nucleotide in the reference genome.\n",
    "We will use zero-based indexing into the reference genome, which is (hopefully!) compatible with Tombo.\n",
    "\n",
    "HDF5 group `/Analyses/RawGenomeCorrected_000/BaseCalled_template/Alignment/` attribute `mapped_start` gives critical information for this.\n",
    "It is the offset between (zero-based) row-index in the Events table and (zero-based) character-index in the reference genome sequence string.\n",
    "(The attribute `mapped_end` gives basically the same info, but I think it's redundant given the length of `events_array`. See the assertion in the following code block.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get parameters to align `events_array` with reference genome\n",
    "with h5py.File(fast5_path, 'r') as file:\n",
    "    mapped_start = file['/Analyses/RawGenomeCorrected_000/BaseCalled_template/Alignment/'].attrs['mapped_start']\n",
    "    mapped_end = file['/Analyses/RawGenomeCorrected_000/BaseCalled_template/Alignment/'].attrs['mapped_end']\n",
    "\n",
    "assert mapped_start + len(events_array) == mapped_end"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The following code block contains an important assertion. It makes sure we are indexing correctly into the reference genome."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create index based on genomic nucleotide position to go with events_table\n",
    "ref_genome_pos = np.arange(mapped_start, mapped_start + len(events_array))\n",
    "\n",
    "# check that for all i the following sentence is true:\n",
    "#     \"The base in row i of events_array equals the base in row ref_genome_pos[i] of reference_genome.\"\n",
    "assert all([reference_genome[ref_genome_pos[i]].upper() == base.upper()\n",
    "            for i, base in enumerate(events_array['base'])])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[(  33, 20)],\n",
       "       [(  34, 17)],\n",
       "       [(  35,  6)],\n",
       "       ...,\n",
       "       [(9163, 27)],\n",
       "       [(9164, 12)],\n",
       "       [(9165, 11)]],\n",
       "      dtype=[('reference_genome_position', '<i8'), ('length', '<u4')])"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# take the index based on reference-genome nucleotide position, and the 'lengths' column, and combine them into one array\n",
    "\n",
    "assert events_array.shape == ref_genome_pos.shape\n",
    "\n",
    "ref_pos_and_length = ( # name means \"reference genome position and length, side by side in the same array\n",
    "    rfn.merge_arrays( # create structured array with two columns...\n",
    "        [\n",
    "            ref_genome_pos, # ...one column from `ref_genome_pos`\n",
    "            events_array['length'] # ...the other column from the `'length'` field of `events_array`\n",
    "        ],\n",
    "        flatten=True\n",
    "    )\n",
    "    .view( # name columns appropriately\n",
    "        np.dtype(\n",
    "            [\n",
    "                 ('reference_genome_position', '<i8'),\n",
    "                 ('length', '<u4')\n",
    "            ]\n",
    "        )\n",
    "    )\n",
    ")\n",
    "\n",
    "# Check again that the reference genome position is correct\n",
    "assert np.array_equal([reference_genome[pos].upper() for pos in ref_pos_and_length['reference_genome_position']], events_array['base'])\n",
    "\n",
    "ref_pos_and_length.reshape(-1,1) # reshape just to make it display in one column"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [],
   "source": [
    "# dump ref_pos_and_length to CSV\n",
    "csv_header = 'reference_genome_position,length\\n'\n",
    "csv_body = ''.join(['{},{}\\n'.format(*row) for row in ref_pos_and_length])\n",
    "csv_text = csv_header + csv_body\n",
    "with open('positions_and_lengths.csv', 'wt') as f:\n",
    "    f.write(csv_text)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## The End?\n",
    "\n",
    "We succeeded at dumping single-read dwell times to a CSV file, indexed by position on the reference genome. Next steps:\n",
    "* Make a script that can do this automatically\n",
    "* Give dwell times in milliseconds, rather than in multiples of the sampling frequency"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2.7 (Conda 5.2) [python/2.7 ]",
   "language": "python",
   "name": "sys_python27conda"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
